public/app/plugins/datasource/loki/components/monaco-query-field/MonacoQueryField.tsx

```diff
@@ -138,6 +150,22 @@ const MonacoQueryField = ({ history, onBlur, onRunQuery, initialValue, datasourc
     editor.onDidChangeModelContent(checkDecorators);
   };
 
+  const onTypeDebounced = debounce(async (query: string) => {
+    if (isValidQuery(query) === false) {
+      return;
+    }
+
+    const timerange = datasource.getTimeRange();
+    const previousTimerange = undefined;
+    const previousQuery = undefined;
+
+    const update = shouldUpdateStats(query, previousQuery, timerange, previousTimerange);
+    if (update) {
+      const stats = await getStats(datasource, query);
+      setQueryStats(stats);
+    }
+  }, 1000);
```

Comment:
Some food for thought after seeing the full PR. I would have considered doing inversion of control here, where we provide prop to execute a debounced `onType` callback passed by the parent, which could be executed here if present.

```suggestion
  const onTypeDebounced = debounce(async (query: string) => {
    if (!onQueryType || isValidQuery(query) === false) {
      return;
    }

    onQueryType();
  }, 1000);
```

Then, in `LokiQueryCodeEditor.tsx`:

```jsx
  <LokiQueryField
        datasource={datasource}
        query={query}
        range={range}
        onRunQuery={onRunQuery}
        onChange={onChange}
        history={history}
        data={data}
        app={app}
        data-testid={testIds.editor}
        onQueryType={() => {
          const timerange = datasource.getTimeRange();
          const previousTimerange = undefined;
          const previousQuery = undefined;

          const update = shouldUpdateStats(query, previousQuery, timerange, previousTimerange);
          if (update) {
            const stats = await getStats(datasource, query);
            setQueryStats(stats);
          }
        }}
      />
```

There are two reasons to consider this. One is that we're providing a less specific component API to any user of `MonacoQueryField`, and we're also not leaking concerns like Query Stats, which are not specific to the Monaco Editor. If tomorrow we need to run `getStats()` and something else "on query type", can just modify the passed callback, and not add another prop to be called inside `MonacoQueryField`. 

Feel free to ask if anything I said is not clear enough.

Reply:
+1 also mentioned it here https://github.com/grafana/grafana/pull/63623#discussion_r1119960063

Reply:
Cool! Apologies for repeating things. I was  checking this PR while we were in the Sync call, in case they asked.